dsp/modules/llm.py

```diff
@@ -0,0 +1,140 @@
+                "kwargs": self.llm_params.to_json(ignore_sensitive=True),
+            }
+        )
+
+        return response
+
+    # TODO: enable caching
+    def request(self, prompt: str, **kwargs) -> ModelResponse:
+        return self.basic_request(prompt, **kwargs)
+
+    def log_usage(self, response: ModelResponse):
+        """Log the total tokens from the OpenAI API response."""
+        usage_data: Usage = response.get("usage")
+        if usage_data:
+            total_tokens = usage_data.get("total_tokens")
+            logger.debug("OpenAI Response Token Usage: %s", total_tokens)
+
+    def filter_only_completed(self, choices: list[Choices]) -> list[Choices]:
+        """Filters out incomplete completions by checking if the finish_reason is not 'length'.
+        Returns the filtered list of choices only if there are any, otherwise returns the original list.
+        """
+        # TODO: This seems provider specific:
+        # Calude uses response.stop_reason == "max_tokens", refer previous code.
+        filtered_choices = [c for c in choices if c.finish_reason != "length"]
+        if len(filtered_choices):
+            return filtered_choices
+        return choices
+
+    def get_text_from_choice(self, choice: Choices) -> str:
+        """Returns the text from the choice."""
+        return choice.message.content
+
+    def transform_choices_to_dspy_model_response(
+        self, choices: list[Choices], add_logprobs: bool = False
+    ) -> list[DSPyModelResponse]:
+        """Transforms the choices to DSPyModelResponse."""
+        dspy_choices: list[DSPyModelResponse] = []
+        for choice in choices:
+            # TODO: ideally we should return the choice object itself, which contains more information.
+            dspy_choices.append(
+                {
+                    "text": self.get_text_from_choice(choice),
+                }
+            )
+            if add_logprobs:
+                # TODO: check if we can strong type this.
+                dspy_choices[-1]["logprobs"] = choice.logprobs if choice.logprobs else None
+
+        return dspy_choices
+
+    def update_messages_with_prompt(self, prompt: str):
+        """Updates the messages with the prompt."""
+        self.llm_params.prompt = prompt
+        if not self.llm_params.messages:
+            self.llm_params.messages = []
+        self.llm_params.messages.append(ChatMessage(role="user", content=prompt))
+
+    def __call__(
+        self,
+        prompt: str,
+        only_completed: bool = True,
+        **kwargs,
+    ) -> list[DSPyModelResponse]:
+        """Retrieves completions from LLM."""
+
+        self.llm_params.only_completed = only_completed
+
+        assert self.llm_params.only_completed, "for now"
+        assert self.llm_params.return_sorted is False, "for now"
+
+        # TODO: I don't really like that the prompt is string instead of messages. We should refactor this.
+        response = self.request(prompt, **kwargs)
```

Comment:
We've been thinking carefully on this - maybe [chatmarkup](https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/chat-markup-language) is a way to letting r-dspy deal in strings and let the litellm wrapper transforms the chat-markup into messages?

Seems like demos and the like could be more effective if dspy was able to leverage the trained assistant/user tokens?

Reply:
This sounds great!